### ISSUE TITLE
Locked LDAP/AD account even though user does not exist in GoCD.

### ISSUE THREAD

**Issue by childsc** [gocd/gocd#6344]
##### Issue Type
- Bug Report

##### Summary

We needed to remove a user from GoCD. But it still exists in AD for other services to use.

Since then we found that when someone tries to log into GoCD with the deleted user via APIs or by the console then there is no pre-check against the user list in GoCD and instead authentication is only on the LDAP server.

This has lead to us having issues where the AD account is being locked because of some pipelines or api calls passing the old username.

Please note that we have also checked with the latest version of GoCD server v19.4.0  and we see the same results.

Also note that making an account in AD not lockable is not an option.

###### Basic environment details

    Go Version: GoCD Version: 18.12.0 (8222-e6778f1cc84bf91e323d337c876046167da36985).
    JAVA Version: OpenJDK 11.0.1
    OS: Linux 3.13.0-147-generic
    Browser vendor and version (if relevant): NA
    LDAP Authentication Plugin for GoCD version 1.0.2-58

##### Steps to Reproduce

1. Create AD user in GoCD "zTest"
2. login to call console or make API call with wrong credentials for "zTest" user  called multiple times, 
3.  Check to see if account is locked

##### Expected Results
AD account is not locked after repeated re-tries

##### Actual Results
AD account is locked

##### Possible Fix
add a check to the plugin or before the plugin is called. as follows:
1st Check user is configured in GoCD
2nd IF LDAP plugin installed
        Then authenticate against LDAP
   ELSE
       Authenticate against local.

**Comment by arvindsv** [comment]
The order of authorization configs in `/go/admin/security/auth_configs` matters. If your file-based password configuration is first in the list then it'll be checked first. However, if the credentials are wrong, I think it'll go on to the other auth configs.

I wonder if the setting in server configuration (`/go/admin/config/server`) which disallows users not in in GoCD from logging in is helpful. It looks like this:

<hr>
<img width="662" alt="2019_05_28_13-25-24" src="https://user-images.githubusercontent.com/172235/58499924-15d21600-814f-11e9-8be9-06c59b5f908e.png">
<hr>

**Comment by arvindsv** [comment]
Changing behavior to not check further if LDAP plugin is installed is probably not going to happen. There are others who're probably depending on the existing behavior. It's allowed to have _multiple_ auth configs.

**Comment by maheshp** [comment]
> I wonder if the setting in server configuration (/go/admin/config/server) which disallows users not in in GoCD from logging in is helpful. It looks like this:

Disallowing users not in GoCD to login might not be useful in this case because this check happens only post authentication with AD and we cannot move this check before authenticating with AD.

**Comment by childsc** [comment]
Hello:

> If your file-based password configuration is first in the list then it'll be checked first.
the LDAP Authorization Configuration is at the top of the list and paswordfile is at the bottom of the list. 

> There are others who're probably depending on the existing behavior. It's allowed to have multiple auth configs.
I guess I can re-work my suggested fix:
1st Check user is configured in GoCD
2nd Authenticate with configured "Authorization Congifuration"

**Comment by maheshp** [comment]
> Disallowing users not in GoCD to login might not be useful in this case because this check happens only post authentication with AD and we cannot move this check before authenticating with AD.

Guess my previous response was incomplete.  Providing more info,

If we need to check for the presence of a user in GoCD before authenticating with AD, we will have to lookup for the user in GoCD based on the username provided in the credentials. It might so happen that the username provided in credentials might not map directly to the username in the DB.

Lets take an example of the [LDAP Authorization Plugin](https://extensions-docs.gocd.org/ldap/current/authorization-configuration/).

**UserLoginFilter** = (|(sAMAccountName={0})(={0})(uid={0}))
**UserNameAttribute** = sAMAccountName

For the above configuration, a user can login to GoCD using either  of  `sAMAccountName`, `mail` or `uid` as username in the credentials. The **UserNameAttribute** in this case `sAMAccountName` will be used to map to the GoCD username field. And this information is provided by the plugin only post the authentication. Hence I think we will not be able to move the presence of user check before making a call to AD.

**Comment by arvindsv** [comment]
I see. So, you're saying someone might be able to login with `user@example.com` but once it goes through the plugin, it might tell GoCD that the username to use is `user1`. Sigh. Ok. I was wondering why it wouldn't be possible to put the user-enabled check before. Might need to do something more explicit (in config) for this.

**Comment by childsc** [comment]
Hello,

Could we not synchronize and store the sAMAccountName, email and uid in Gocd as well. I think this is how other applications can do it?

**Comment by arvindsv** [comment]
@childsc No. We cannot. What @maheshp provided was just an example. That field is fully configurable and people use it in all kinds of ways. I know of weird Active Directory specific queries that people have used to do all kinds of things there. Bringing that into GoCD is not right.

**Comment by childsc** [comment]
@arvindsv and @maheshp thanks for your help with this.

As we saying that this is working by design?
So are we saying that there is not really anything that can be done to prevent my accounts getting locked? 

What would you consider my options to be?

**Comment by arvindsv** [comment]
@childsc Currently, it is working as designed. _However_ that does not mean that the case you've brought up is invalid. :)

@maheshp and I have been talking and what we know (apart from what is mentioned earlier) is:

- If there are multiple auth configs, there is no way for one auth config to short-circuit all the others. In the example below, even if the first one fails, it'll continue to the second.

- In the [plugin API](https://plugin-api.gocd.org/current/authorization/#authenticate-user), there is no way for a plugin to tell GoCD server the difference between: "I have no idea about this user" vs. "This user's authentication failed".

- If we did have that ability to differentiate it, we could provide an option at the `authConfig` level, which would short-circuit. Something like:

```xml
<authConfigs>
    <authConfig id="file-auth-config" continueOnAuthenticationFailure="false" pluginId="cd.go.authentication.passwordfile">
        ...
    </authConfig>

    <authConfig id="ldap-auth-config" continueOnAuthenticationFailure="false" pluginId="cd.go.authentication.ldap" continueOnAuthenticationFailure="true">
        ...
    </authConfig>

    <authConfig id="another-file-auth-config" continueOnAuthenticationFailure="false" pluginId="cd.go.authentication.passwordfile">
        ...
    </authConfig>
</authConfigs>
```

This involves changes to the config (above), the plugin API (mentioned earlier), the [GoCD auth config API](https://api.gocd.org/current/#authorization-configuration), the [plugins](https://github.com/gocd/gocd-filebased-authentication-plugin) [themselves](https://github.com/gocd/gocd-ldap-authentication-plugin) to differentiate between the two cases mentioned above and finally to GoCD's auth mechanisms. It's not small. So, there is no short-term or medium-term fix, unless you want to fork the LDAP plugin, which is probably overkill.

Hmm. It might be possible to blacklist based on inputs to the plugin actually. So, in its config, you'd have a list of users who shouldn't be allowed to log in.

**Comment by childsc** [comment]
@arvindsv  I see that you have mentioned in that this issue would not be fixed in #6571 .

Where does this leave us? :)

**Comment by maheshp** [comment]
> @arvindsv I see that you have mentioned in that this issue would not be fixed in #6571 .
> Where does this leave us? :)

@childsc issue https://github.com/gocd/gocd/issues/6571 though dealing with Authentication/Authorization is different from the issue raised by you. We still need to address this issue maybe as part of #6571 but the considering the other things being worked on it might not be picked up immediately.
